File: reference-code/src/main/java/createcommands/CommandTrees.java
package createcommands;

import dev.jorel.commandapi.CommandAPI;
import dev.jorel.commandapi.CommandTree;
import dev.jorel.commandapi.arguments.DoubleArgument;
import dev.jorel.commandapi.arguments.GreedyStringArgument;
import dev.jorel.commandapi.arguments.IntegerArgument;
import dev.jorel.commandapi.arguments.LiteralArgument;
import dev.jorel.commandapi.arguments.PlayerArgument;
import dev.jorel.commandapi.arguments.StringArgument;
import dev.jorel.commandapi.exceptions.WrapperCommandSyntaxException;
import org.bukkit.block.Block;
import org.bukkit.block.Sign;
import org.bukkit.entity.Player;
import org.bukkit.metadata.FixedMetadataValue;
import org.bukkit.plugin.java.JavaPlugin;
import org.jetbrains.annotations.NotNull;

class CommandTrees {
    {
        // #region commandTreesExample
        new CommandTree("sayhi")
            .executes((sender, args) -> {
                sender.sendMessage("Hi!");
            })
            .then(new PlayerArgument("target")
                .executes((sender, args) -> {
                    Player target = (Player) args.get("target");
                    target.sendMessage("Hi");
                }))
            .register();
        // #endregion commandTreesExample

        // #region legacyLargeArgumentsExample
        new CommandTree("example")
            .then(new LiteralArgument("arg1")
                .then(new StringArgument("arg2")
                    .then(new StringArgument("arg3")
                        .then(new DoubleArgument("arg4", 0)
                            .then(new StringArgument("arg5")
                                .executes((sender, args) -> {
                                    // your code here
                                }))))))
            .register();
        // #endregion legacyLargeArgumentsExample

        // #region nestedLargeArgumentsExample
        new CommandTree("example")
            .thenNested(
                new LiteralArgument("arg1"),
                new StringArgument("arg2"),
                new StringArgument("arg3"),
                new DoubleArgument("arg4", 0),
                new StringArgument("arg5")
                    .executes((sender, args) -> {
                        // your code here
                    })
            ).register();
        // #endregion nestedLargeArgumentsExample

        new JavaPlugin() {
            @Override
            public void onEnable() {
                // #region signeditCommandExample
                new CommandTree("signedit")
                    .then(new LiteralArgument("set")
                        .then(new IntegerArgument("line_number", 1, 4)
                            .then(new GreedyStringArgument("text")
                                .executesPlayer((player, args) -> {
                                    // /signedit set <line_number> <text>
                                    Sign sign = getTargetSign(player);
                                    int lineNumber = (int) args.get("line_number");
                                    String text = (String) args.get("text");
                                    sign.setLine(lineNumber - 1, text);
                                    sign.update(true);
                                }))))
                    .then(new LiteralArgument("clear")
                        .then(new IntegerArgument("line_number", 1, 4)
                            .executesPlayer((player, args) -> {
                                // /signedit clear <line_number>
                                Sign sign = getTargetSign(player);
                                int lineNumber = (int) args.get("line_number");
                                sign.setLine(lineNumber - 1, "");
                                sign.update(true);
                            })))
                    .then(new LiteralArgument("copy")
                        .then(new IntegerArgument("line_number", 1, 4)
                            .executesPlayer((player, args) -> {
                                // /signedit copy <line_number>
                                Sign sign = getTargetSign(player);
                                int lineNumber = (int) args.get("line_number");
                                player.setMetadata("copied_sign_text", new FixedMetadataValue(this, sign.getLine(lineNumber - 1)));
                            })))
                    .then(new LiteralArgument("paste")
                        .then(new IntegerArgument("line_number", 1, 4)
                            .executesPlayer((player, args) -> {
                                // /signedit copy <line_number>
                                Sign sign = getTargetSign(player);
                                int lineNumber = (int) args.get("line_number");
                                sign.setLine(lineNumber - 1, player.getMetadata("copied_sign_text").get(0).asString());
                                sign.update(true);
                            })))
                    .register();
                // #endregion signeditCommandExample
            }
        };
    }

    public Sign getTargetSign(@NotNull Player player) throws WrapperCommandSyntaxException {
        Block block = player.getTargetBlock(null, 256);
        if (block.getState() instanceof Sign sign) {
            return sign;
        } else {
            throw CommandAPI.failWithString("You're not looking at a sign!");
        }
    }
}

File: reference-code/src/main/kotlin/createcommands/CommandTrees.kt
package createcommands

import dev.jorel.commandapi.CommandAPI
import dev.jorel.commandapi.CommandTree
import dev.jorel.commandapi.arguments.*
import dev.jorel.commandapi.executors.CommandExecutor
import dev.jorel.commandapi.executors.PlayerCommandExecutor
import dev.jorel.commandapi.kotlindsl.*
import org.bukkit.block.Sign
import org.bukkit.entity.Player
import org.bukkit.metadata.FixedMetadataValue
import org.bukkit.plugin.java.JavaPlugin

fun commandTrees() {
    // #region commandTreesExample
    CommandTree("sayhi")
        .executes(CommandExecutor { sender, _ ->
            sender.sendMessage("Hi!")
        })
        .then(
            PlayerArgument("target")
                .executes(CommandExecutor { _, args ->
                    val target = args["target"] as Player
                    target.sendMessage("Hi")
                })
        )
        .register()
    // #endregion commandTreesExample

    // @formatter:off
    // #region legacyLargeArgumentsExample
    CommandTree("example")
        .then(LiteralArgument("arg1")
            .then(StringArgument("arg2")
                .then(StringArgument("arg3")
                    .then(DoubleArgument("arg4", 0.0)
                        .then(StringArgument("arg5")
                            .executes(CommandExecutor { sender, _ ->
                                // your code here
                            }))))))
        .register()
    // #endregion legacyLargeArgumentsExample
    // @formatter:on

    // #region nestedLargeArgumentsExample
    CommandTree("example")
        .thenNested(
            LiteralArgument("arg1"),
            StringArgument("arg2"),
            StringArgument("arg3"),
            DoubleArgument("arg4", 0.0),
            StringArgument("arg5")
                .executes(CommandExecutor { sender, _ ->
                    // your code here
                })
        ).register()
    // #endregion nestedLargeArgumentsExample

    object : JavaPlugin() {
        override fun onEnable() {
            // @formatter:off
            // #region signeditCommandExample
            CommandTree("signedit")
                .then(LiteralArgument("set")
                    .then(IntegerArgument("line_number", 1, 4)
                        .then(GreedyStringArgument("text")
                            .executesPlayer(PlayerCommandExecutor { player, args ->
                                // /signedit set <line_number> <text>
                                val sign: Sign = getTargetSign(player)
                                val line_number = args["line_number"] as Int
                                val text = args["text"] as String
                                sign.setLine(line_number - 1, text)
                                sign.update(true)
                            }))))
                .then(LiteralArgument("clear")
                    .then(IntegerArgument("line_number", 1, 4)
                        .executesPlayer(PlayerCommandExecutor { player, args ->
                            // /signedit clear <line_number>
                            val sign: Sign = getTargetSign(player)
                            val line_number = args["line_number"] as Int
                            sign.setLine(line_number - 1, "")
                            sign.update(true)
                        })))
                .then(LiteralArgument("copy")
                    .then(IntegerArgument("line_number", 1, 4)
                        .executesPlayer(PlayerCommandExecutor { player, args ->
                            // /signedit copy <line_number>
                            val sign: Sign = getTargetSign(player)
                            val line_number = args["line_number"] as Int
                            player.setMetadata("copied_sign_text", FixedMetadataValue(this, sign.getLine(line_number - 1)))
                        })))
                .then(LiteralArgument("paste")
                    .then(IntegerArgument("line_number", 1, 4)
                        .executesPlayer(PlayerCommandExecutor { player, args ->
                            // /signedit copy <line_number>
                            val sign: Sign = getTargetSign(player)
                            val line_number = args["line_number"] as Int
                            sign.setLine(line_number - 1, player.getMetadata("copied_sign_text")[0].asString())
                            sign.update(true)
                        })))
                .register()
            // #endregion signeditCommandExample
            // @formatter:on
        }

        private fun getTargetSign(player: Player): Sign = player.getTargetBlock(null, 256).state as? Sign ?: throw CommandAPI.failWithString("You're not looking at a sign!")
    }
}

fun commandTreesDSL() {
    // #region legacyLargeArgumentsExampleDSL
    commandTree("example") {
        literalArgument("arg1") {
            stringArgument("arg2") {
                stringArgument("arg3") {
                    doubleArgument("arg4", 0.0) {
                        stringArgument("arg5") {
                            anyExecutor { sender, _ ->
                                // your code here
                            }
                        }
                    }
                }
            }
        }
    }
    // #endregion legacyLargeArgumentsExampleDSL

    // #region nestedLargeArgumentsExampleDSL
    commandTree("example") {
        nested {
            literalArgument("arg1")
            stringArgument("arg2")
            stringArgument("arg3")
            doubleArgument("arg4", 0.0)
            stringArgument("arg5") {
                anyExecutor { sender, _ ->
                    // your code here
                }
            }
        }
    }
    // or
    commandTree("example") {
        nestedArguments(
            LiteralArgument("arg1"),
            StringArgument("arg2"),
            StringArgument("arg3"),
            DoubleArgument("arg4", 0.0),
            StringArgument("arg5")
        ) {
            anyExecutor { sender, _ ->
                // your code here
            }
        }
    }
    // #endregion nestedLargeArgumentsExampleDSL
}